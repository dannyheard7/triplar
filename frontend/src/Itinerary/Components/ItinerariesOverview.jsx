import React from "react";
import {connect} from "react-redux";
import {getTripItineraries} from "../utils/actions";
import {Link} from "react-router-dom";
import Moment from "moment";
import {faPlus} from "@fortawesome/free-solid-svg-icons";
import {FontAwesomeIcon} from "@fortawesome/react-fontawesome";

import 'Itinerary/styles/itineraries.css';
import LocationAddContainer from "./LocationAdd";


export class ItinerariesOverview extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            showLocationAdd: false
        };

        this.onClick = this.onClick.bind(this);
    }

    onClick() {
        this.setState({showLocationAdd: !this.state.showLocationAdd});
    }

    componentDidMount() {
        this.props.dispatch(getTripItineraries(this.props.match.params.id));
    }

    render() {
        const trip = this.props.trip;
        const locations = this.props.locations;

        return (
            <div className="itineraries-overview">
                <h2>Locations</h2>
                <div className="itineraries-list">
                    {locations.map((location) =>
                        <div className="card" key={location.id}>
                            <Link to={`/trips/${trip.id}/itinerary/${location.id}`}>
                                <h3 className="card-title">{location.city.name}{location.city.country && ", " + location.city.country.name}</h3>
                            </Link>
                            <p className="card-text">
                                {Moment(location.startDate).format('Do MMMM')} - {Moment(location.endDate).format('Do MMMM')}
                            </p>
                        </div>
                    )}
                    <div className="card">
                        {this.state.showLocationAdd ?
                            <LocationAddContainer trip={trip}/> :
                            <button className="btn btn-primary" onClick={this.onClick}><FontAwesomeIcon icon={faPlus}/></button>
                        }
                    </div>
                </div>
            </div>
        )
    }
}

ItinerariesOverview.defaultProps = {
    trip: {},
    locations: []
};

function mapStateToProps(state, ownProps) {
    return {
        trip: state.trips.trips.find(x => x.id === ownProps.match.params.id),
        locations: state.itineraries.locationItineraries.filter(x => x.trip.id === ownProps.match.params.id)
    };
}

export default connect(mapStateToProps)(ItinerariesOverview)
