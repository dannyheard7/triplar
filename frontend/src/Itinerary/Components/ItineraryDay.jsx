import React from "react";
import Moment from 'moment';

export default function ItineraryDay({day}) {
    day = Moment(day);

    return (
        <p>
            {day.format("dddd Do MMMM")}
        </p>
    );
}
