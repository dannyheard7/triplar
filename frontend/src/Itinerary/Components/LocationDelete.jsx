import React from "react";

import Modal from 'react-responsive-modal';
import {connect} from "react-redux";
import {deleteTripLocation} from "../utils/actions";
import LoadingIndicator from "../../App/Components/LoadingIndicator";

export class LocationDelete extends React.Component {
    constructor(props) {
        super(props);

        this.onDelete = this.onDelete.bind(this);
        this.closeModal = this.closeModal.bind(this);
    }

    closeModal() {
        // Should do this or rewrite path?
        this.props.history.goBack();
    }

    onDelete(e) {
        this.props.dispatch(deleteTripLocation(this.props.itinerary.id, this.props.match.params.tripId));
    };

    render() {
        if(this.props.itinerary) {
            return (
                <Modal open={true} onClose={this.closeModal} contentLabel={this.props.itinerary.city.name} >
                    <h2>Remove {this.props.itinerary.city.name}?</h2>

                    <p>Are you sure you want to remove the location '{this.props.itinerary.city.name}'?</p>

                    <button onClick={this.closeModal}>Close</button>
                    <button className="btn-danger" onClick={this.onDelete}>Remove</button>
                </Modal>
            );
        } else {
            return <LoadingIndicator/>
        }
    }
}

function mapStateToProps(state, ownProps) {
  return {
      itinerary: state.itineraries.locationItineraries.find(x => x.id === ownProps.match.params.itineraryId),
  };
}

export default connect(mapStateToProps)(LocationDelete)