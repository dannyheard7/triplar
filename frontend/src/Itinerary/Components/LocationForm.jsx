import React from "react";
import NonFieldErrors from "../../Forms/Components/NonFieldErrors";
import CitySearchFieldContainer from "../../Forms/Containers/CitySearchFieldContainer";
import DateFieldGroup from "../../Forms/Components/DateFieldGroup";
import moment from "moment";

moment.locale('en');

export default class LocationForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            arrivalDate: props.location.arrivalDate ? moment(props.location.arrivalDate) : moment(),
            departureDate: props.location.departureDate ? moment(props.location.departureDate) : moment()
        };

        this.onDateChange = this.onDateChange.bind(this);
        this.calculateEndDate = this.calculateEndDate.bind(this);
    }

    componentDidUpdate(prevProps, prevState) {
        if(prevProps.location.arrivalDate !== this.props.location.arrivalDate || prevProps.location.departureDate !== this.props.location.departureDate) {
            this.setState( {
                arrivalDate: moment(this.props.location.arrivalDate),
                departureDate: moment(this.props.location.departureDate)
            });
        }
    }

    onDateChange(value, date) {
        this.setState({
            [value]: moment(date, "DD/MM/YYYY")
        });
    }

    calculateEndDate(startDate) {
        const location = this.props.location;

        if(this.state.departureDate && this.state.departureDate.format("YYYY-MM-DD") !== location.departureDate) {
            return this.state.departureDate;
        } else {
            return startDate.clone().add(1, 'days');
        }
    }

    render() {
        const fieldErrors = this.props.fieldErrors;
        const itinerary = (this.props.location) ? this.props.location : {};
        const location = itinerary.city ? itinerary.city.name_std + ", " + itinerary.city.country : "";

        const arrivalDate = this.state.arrivalDate;
        const departureDate = this.calculateEndDate(this.state.arrivalDate);

        return (
            <form onSubmit={this.props.onSubmit}>
                <NonFieldErrors errors={this.props.nonFieldErrors}/>
                <CitySearchFieldContainer errors={fieldErrors.locations} value={location} name="city"/>
                <DateFieldGroup errors={fieldErrors.arrivalDate} label="Arrival" name="arrivalDate"
                                value={arrivalDate.format("DD/MM/YYYY")} required={true}
                                onChange={(date) => this.onDateChange("arrivalDate", date)}/>
                <DateFieldGroup errors={fieldErrors.departureDate} label="Departure" name="departureDate"
                                value={departureDate.format("DD/MM/YYYY")} required={true}
                                onChange={(date) => this.onDateChange("departureDate", date)}/>
                <button type="submit" className="btn btn-primary">{this.props.submitLabel}</button>
            </form>
        );
    }
}
