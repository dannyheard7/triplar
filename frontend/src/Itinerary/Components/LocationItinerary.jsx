import React from "react";
import Moment from 'moment';
import {daysBetweenDates} from 'Trips/utils/utils.js';
import PlacesSearchContainer from "Places/Containers/PlacesSearchContainer";
import ItineraryDayContainer from "Itinerary/Containers/ItineraryDayContainer";
import {connect} from "react-redux";
import {getItineraryDayDetail} from "../utils/actions";
import MarkerMap from "../../Maps/Components/MarkerMap";


export class LocationItinerary extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            showMap: false,
            selectedDate: props.itinerary.startDate,
        };

        this.onClick = this.onClick.bind(this);
        this.onChangeSelectedDay = this.onChangeSelectedDay.bind(this);
    }

    componentDidMount() {
        this.props.dispatch(getItineraryDayDetail(this.props.itinerary.id, this.state.selectedDate));
    }

    onClick() {
        this.setState({showMap: !this.state.showMap});
    }

    onChangeSelectedDay(event) {
        this.setState({selectedDate: event.target.value});
        this.componentDidMount();
    }

    render () {
        const itinerary = this.props.itinerary;
        const days = daysBetweenDates(itinerary.startDate, itinerary.endDate);

        return (
            <div className="destination-itinerary">
                <div className="destination-plan">
                    <h3 onClick={this.onClick}>{itinerary.city.name}</h3>
                    <p>{Moment(itinerary.startDate).format('dddd Do MMMM') } - {Moment(itinerary.endDate).format('dddd Do MMMM') }
                        <select onChange={this.onChangeSelectedDay}>
                            {days.map((day) => <option key={day} value={day.format("YYYY-MM-DD")}>{day.format('dddd Do MMMM')}</option>)}
                        </select>
                    </p>
                    <ItineraryDayContainer day={this.state.selectedDate} itinerary={itinerary} key={this.state.selectedDate}
                        path={`/trips/${this.props.tripId}`} />
                </div>
                {itinerary.city.location && <PlacesSearchContainer city={itinerary.city} path={`/trips/${this.props.tripId}`}/>}
            </div>
        )
    }
}

LocationItinerary.defaultProps = {
    itinerary: {
        city: {
            location: {
                lat: 0,
                lng: 0
            }
        }
    }
};

function mapStateToProps(state, ownProps) {
  return {
      itinerary: state.itineraries.locationItineraries.find(x => x.id === ownProps.location.id)
  };
}

export default connect(mapStateToProps)(LocationItinerary)
