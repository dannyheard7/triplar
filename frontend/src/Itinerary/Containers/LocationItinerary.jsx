import React from "react";
import Moment from 'moment';
import {connect} from "react-redux";

import {daysBetweenDates} from '../utils/utils.js';
import PlacesSearchContainer from "../../Places/Containers/PlacesSearchContainer";
import ItineraryDayContainer from "./ItineraryDayContainer";
import {addItemToItineraryDay, getItineraryDayDetail, removeItemFromItineraryDay} from "../utils/actions";
import {Helmet} from "react-helmet";
import {DragDropContext} from "react-beautiful-dnd";


export class LocationItinerary extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            selectedDate: props.itinerary.startDate,
            dragging: false,
        };

        this.onChangeSelectedDay = this.onChangeSelectedDay.bind(this);
        this.onDragEnd = this.onDragEnd.bind(this);
        this.onDragStart = this.onDragStart.bind(this);
    }

    componentDidMount() {
        this.props.dispatch(getItineraryDayDetail(this.props.itinerary.id, this.state.selectedDate));
    }

    componentDidUpdate(prevProps) {
        if (this.props.itinerary.id !== prevProps.itinerary.id) {
            this.setState({selectedDate: this.props.itinerary.startDate}, this.componentDidMount);
        }
    }

    onChangeSelectedDay(event) {
        this.setState({selectedDate: event.target.value}, this.componentDidMount);
    }

    onDragEnd(result) {
        this.setState({dragging: false});
        if (!result.destination) return;

        const itineraryId = this.props.itinerary.id;
        let placeId = result.draggableId;
        if(result.source.droppableId === "itinerary-day-droppable") {
            placeId = result.draggableId.slice(0, -itineraryId.toString().length);
        }

        if (result.destination.droppableId === result.source.droppableId) {
            //  reorder
            return;
        }

        const selectedDate = Moment(this.state.selectedDate).format("YYYY-MM-DDz");

        if (result.destination.droppableId === "placeSearchContainer") {
            this.props.dispatch(removeItemFromItineraryDay(itineraryId, placeId, selectedDate));
        } else if (!this.props.itineraryItems.filter(x => x.date === selectedDate).map(x => x.place).includes(placeId)) {
            this.props.dispatch(addItemToItineraryDay(itineraryId, placeId, selectedDate, result.destination.index));
        }
    }

    onDragStart() {
        this.setState({dragging: true});
    }

    render() {
        const {trip, itinerary} = this.props;
        const tripId = trip ? trip.id : this.props.match.params.tripId;
        const path = `/trips/${tripId}/itinerary/${itinerary.id}`;

        const days = daysBetweenDates(itinerary.startDate, itinerary.endDate);

        return (
            <DragDropContext onDragEnd={this.onDragEnd} onDragStart={this.onDragStart}>
                <div>
                    <Helmet>
                        <title>{trip.name + ": " + itinerary.city.name + " | Triplar"}</title>
                    </Helmet>
                    <div>
                        <div className="row">
                            <h3>{itinerary.city.name}</h3>
                            <select onChange={this.onChangeSelectedDay}>
                                {days.map((day) =>
                                    <option key={day} value={day.format("YYYY-MM-DD")}>{day.format('dddd Do MMMM')}</option>
                                )}
                            </select>
                        </div>
                        <ItineraryDayContainer day={this.state.selectedDate} itinerary={itinerary} key={this.state.selectedDate}
                                               path={path} dragging={this.state.dragging}/>
                    </div>
                    <PlacesSearchContainer city={itinerary.city} path={path}/>
                </div>
            </DragDropContext>
        )
    }
}

LocationItinerary.defaultProps = {
    itinerary: {city: {}},
    trip: {}
};

function mapStateToProps(state, ownProps) {
    const itineraryId = ownProps.match.params.itineraryId;
    return {
        itineraryItems: state.itineraries.itineraryItems.filter(x => x.itineraryId === itineraryId),
        itinerary: state.itineraries.locationItineraries.find(x => x.id === itineraryId),
        trip: state.trips.trips.find(x => x.id === ownProps.match.params.tripId)
    };
}

export default connect(mapStateToProps)(LocationItinerary)
