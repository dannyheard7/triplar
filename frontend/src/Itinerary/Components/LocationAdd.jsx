import React from "react";
import LocationForm from "./LocationForm";
import ReduxFormContainer from "../../Forms/Containers/ReduxFormContainer";
import {addTripLocation} from "../utils/actions";
import {connect} from "react-redux";

export class LocationAdd extends React.Component {
    curryAction = (action, id) => (object) => action(id, object);

    render () {
        const errors = this.props.errors ? this.props.errors : [];

        const location = {arrivalDate: this.props.trip.startDate, department: this.props.trip.endDate};

        return (
            <ReduxFormContainer action={this.curryAction(addTripLocation, this.props.trip.id)} fieldErrors={errors} >
                <LocationForm submitLabel="Add" location={location}/>
            </ReduxFormContainer>
        )
    }
}

const mapStateToProps = (state) => ({
    errors: state.errors.ADD_TRIP_LOCATION
});
export default connect(mapStateToProps)(LocationAdd)