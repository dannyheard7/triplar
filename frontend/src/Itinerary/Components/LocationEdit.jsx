import React from "react";
import LocationForm from "./LocationForm";
import ReduxFormContainer from "../../Forms/Containers/ReduxFormContainer";
import {addTripLocation} from "../utils/actions";
import {connect} from "react-redux";

export class LocationAdd extends React.Component {
    curryAction = (action, id) => (object) => action(id, object);

    render () {
        return (
            <ReduxFormContainer action={this.curryAction(addTripLocation, this.props.trip.id)} errors={this.props.errors} >
                <LocationForm submitLabel="Add" location={this.props.trip}/>
            </ReduxFormContainer>
        )
    }
}

const mapStateToProps = (state) => {return {errors: state.itineraries.errors}};
export default connect(mapStateToProps)(LocationAdd)