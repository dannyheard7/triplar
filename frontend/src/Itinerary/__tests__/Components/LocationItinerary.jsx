import React from "react";

import {LocationItinerary} from "Itinerary/Components/LocationItinerary";
import ShallowRenderer from "react-test-renderer/shallow";
import {shallow} from "enzyme/build";
import MarkerMap from "../../../Maps/Components/MarkerMap";

const renderer = new ShallowRenderer();
jest.mock('Itinerary/utils/itinerary.api.js');

describe('<LocationItinerary />', () => {
    const props = {
        itinerary: {
            trip: {
                id: 1
            },
            city: {
                name: "Bristol",
                country: "United Kingdom",
                location: {
                    lat: 0.0,
                    lng: 0.0
                }
            },
            startDate: "2018-08-10T00:00:00.000Z",  // If the time isn't specified, the generated snapshot assumes it is the day before?
            endDate: "2018-08-12T00:00:00.000Z",
        },
        match: {
            params: {
                tripId: null
            }
        },
        dispatch: jest.fn()
    };

    test('<LocationItinerary /> renders correctly', () => {
        const result = renderer.render(<LocationItinerary {...props} />);
        expect(result).toMatchSnapshot();
    });
});