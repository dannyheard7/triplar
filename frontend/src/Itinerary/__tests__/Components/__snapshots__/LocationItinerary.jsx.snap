// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LocationItinerary /> <LocationItinerary /> renders correctly 1`] = `
<DragDropContext
  onDragEnd={[Function]}
  onDragStart={[Function]}
>
  <div
    className="destination-itinerary"
  >
    <HelmetWrapper
      defer={true}
      encodeSpecialCharacters={true}
    >
      <title>
        undefined: Bristol | Triplar
      </title>
    </HelmetWrapper>
    <div
      className="destination-plan"
    >
      <h3>
        Bristol
      </h3>
      <p>
        Friday 10th August
         - 
        Sunday 12th August
        <select
          onChange={[Function]}
        >
          <option
            value="2018-08-10"
          >
            Friday 10th August
          </option>
          <option
            value="2018-08-11"
          >
            Saturday 11th August
          </option>
          <option
            value="2018-08-12"
          >
            Sunday 12th August
          </option>
        </select>
      </p>
      <Connect(ItineraryDayContainer)
        day="2018-08-10T00:00:00.000Z"
        dragging={false}
        itinerary={
          Object {
            "city": Object {
              "country": "United Kingdom",
              "location": Object {
                "lat": 0,
                "lng": 0,
              },
              "name": "Bristol",
            },
            "endDate": "2018-08-12T00:00:00.000Z",
            "startDate": "2018-08-10T00:00:00.000Z",
            "trip": Object {
              "id": 1,
            },
          }
        }
        path="/trips/1/itinerary/undefined"
      />
    </div>
    <Connect(PlacesSearchContainer)
      city={
        Object {
          "country": "United Kingdom",
          "location": Object {
            "lat": 0,
            "lng": 0,
          },
          "name": "Bristol",
        }
      }
      path="/trips/1/itinerary/undefined"
    />
  </div>
</DragDropContext>
`;
